### ISSUE TITLE
Call Stack problems

### ISSUE THREAD

**Issue by Yenza** [chandlerprall/ThreeCSG#23]
Using three.min.js rev 73

Method used:
var roomBSP = new ThreeBSP(_RoomShape); // HUGE SHAPE

Info gathered from a TryCatch:
"number : -2146828260"
"Error: Out of stack space
at ThreeBSP.Vertex.prototype.dot (*/ThreeCSG.js)
at ThreeBSP.Polygon.prototype.classifyVertex (*/ThreeCSG.js) 
at ThreeBSP.Polygon.prototype.classifySide (*/ThreeCSG.js) 
at ThreeBSP.Polygon.prototype.splitPolygon (*/ThreeCSG.js) 
at ThreeBSP.Node (*/ThreeCSG.js) 
at ThreeBSP.Node (*/ThreeCSG.js) 
at ThreeBSP.Node (*/ThreeCSG.js) 
at ThreeBSP.Node (*/ThreeCSG.js) 
at ThreeBSP.Node (*/ThreeCSG.js) 
at ThreeBSP.Node (*/ThreeCSG.js)"

Using the OUTDATED csg-wrapper from kraag22/csg-wrapper it doesn't happen but his wrapper is way outdated and needs an update.

Might this occur when using high values/numbers?

Another example:
var intersect = shape1.intersect(shape2); // Small shapes

Info gathered from a TryCatch:
"number : -2146828260"
-:-

**Comment by Wilt** [comment]
I have an updated fork [here](https://github.com/Wilt/ThreeCSG)

**Comment by Yenza** [comment]
Your fork is exactly the same except missing the face.vertexNormal fix

**Comment by Wilt** [comment]
Owh, sorry. I thought that there were more fixes in my fork...

**Comment by Wilt** [comment]
I provided the vertex normal fix... 
But I accidently pointed to the master, and that fix was [in this patch1 branch.](https://github.com/Wilt/ThreeCSG/blob/patch-1/ThreeCSG.js)

**Comment by Yenza** [comment]
Your branch is fully up-to-date with Chandlers and you both provide exact same ThreeCSG.js files.
I've just compared with a SVN program. No changes.

**Comment by Wilt** [comment]
@Yenza yes, you are right. [he recently merged my patch1...](https://github.com/chandlerprall/ThreeCSG/commits?author=Wilt)

**Comment by chandlerprall** [comment]
Maximum call stack happens with complex geometry for two reasons:
1) ThreeCSG [doesn't use any heuristic](https://github.com/chandlerprall/ThreeCSG/blob/master/ThreeCSG.js#L444) for picking a good divider
2) Even if a great divider is selected it is still possible to reach maximum recursion. The proper fix would be to change the recursion into one function that keeps a local stack queue and while() loops over it.

Solving either of these would gain large speed improvements, especially for complex geometries. There are many white papers / thesis on CSG which provide ideas for selecting a good divider. Unfortunately I won't have time for a while to actively look into implementing.

**Comment by Yenza** [comment]
@Wilt Do you think you could try to fix this issue?

@chandlerprall But it isn't that very complex! It's two round shapes! It's two identical "almost" circle-wise shapes on the same height. The thing is that if I move any of them like 1 unit in any direction.. then it works! I have an image of the red area when using the intersect method that causes an error:
http://i.imgur.com/NW0AOUU.png

Also this problem doesn't seem to be occurring when you use the old csg wrapper but that might just be because it's heavily outdated.

**Comment by Yenza** [comment]
So this is what I did to fix this for now... so there isn't any max call stack
```
var intersection;
try {
    // If it doesn't work the first time...
    intersection= shapeBSP.intersect(shapeBSP_2);
}
catch (e) {
    // Then try again..
    var newShape = shapeBSP_2.toMesh(_Material);
    newShape.position.x += 1;
    newShape = new ThreeBSP(newShape );
    intersection= shapeBSP.intersect(newShape);
    // console.warn("Works!");
}
```

**Comment by jgphilpott** [comment]
FYI [this new CSG repo](https://github.com/giladdarshan/OctreeCSG) using [Octree](https://en.wikipedia.org/wiki/Octree) data structure overcomes a lot of the performance issues with ThreeCSG and other CSG tools. See [this SO thread](https://stackoverflow.com/a/76999588/1544937) for up to date information.
